backend/parsers/windmill-parser-graphql/src/lib.rs

```diff
@@ -7,53 +7,55 @@ use serde_json::json;
 use windmill_parser::{Arg, MainArgSignature, Typ};
 
 pub fn parse_graphql_sig(code: &str) -> anyhow::Result<MainArgSignature> {
-    let parsed = parse_graphql_file(&code)?;
-    if let Some(x) = parsed {
-        let args = x;
-        Ok(MainArgSignature { star_args: false, star_kwargs: false, args, no_main_func: None })
-    } else {
-        Err(anyhow!("Error parsing sql".to_string()))
-    }
+    parse_graphql_file(&code)
+        .map(|args| MainArgSignature {
+            star_args: false,
+            star_kwargs: false,
+            args,
+            no_main_func: None,
+        })
+        .ok_or_else(|| anyhow!("Error parsing sql".to_string()))
 }
 
 lazy_static::lazy_static! {
     static ref RE_ARG_GRAPHQL: Regex = Regex::new(r#"\$(\w+)\s*:\s*(?:(\w+)!?|\[(\w+)!?\])!?\s*(?:=\s*(\w+)\s*)?"#).unwrap();
 }
 
-fn parse_graphql_file(code: &str) -> anyhow::Result<Option<Vec<Arg>>> {
-    let mut args: Vec<Arg> = vec![];
-
-    for cap in RE_ARG_GRAPHQL.captures_iter(code) {
-        let name = cap.get(1).map(|x| x.as_str().to_string()).unwrap();
-        let mut typ = cap.get(2).map(|x| x.as_str().to_string());
-
-        let parsed_typ = if typ.is_none() {
-            let inner_typ = cap.get(3).map(|x| x.as_str().to_string());
-            typ = inner_typ.clone().map(|x| format!("[{}]", x.to_string()));
-            Typ::List(Box::new(parse_graphql_typ(inner_typ.unwrap().as_str())))
-        } else {
-            parse_graphql_typ(typ.clone().unwrap().as_str())
-        };
-
-        let default = cap.get(4).map(|x| x.as_str().to_string());
-
-        let has_default = default.is_some();
-
-        let parsed_default = default.and_then(|x| match parsed_typ {
-            Typ::Int => x.parse::<i64>().ok().map(|x| json!(x)),
-            Typ::Float => x.parse::<f64>().ok().map(|x| json!(x)),
-            _ => Some(json!(x)),
-        });
-        args.push(Arg {
-            name,
-            typ: parsed_typ,
-            default: parsed_default,
-            otyp: Some(typ.unwrap()),
-            has_default,
-        });
-    }
-
-    Ok(Some(args))
+fn parse_graphql_file(code: &str) -> Option<Vec<Arg>> {
+    RE_ARG_GRAPHQL
+        .captures_iter(code)
+        .filter_map(|cap| Some(parse_arg_cap(cap)))
+        .collect()
+}
+fn parse_arg_cap(cap: regex::Captures) -> Option<Arg> {
```

Comment:
Why does this function return an option?

Reply:
Seems that I misunderstood , initial implementation so i thought wrapping option into Result was extra at that point .
- Reverted back to initial signature `anyhow::Result<Option<Arg>>`
- also I figured simple macro capturing cleans up a lot of code if we pre extract it all into 4x tuple
- I also noticed that Err should probable state "Error parsing gql" instead of "Error parsing sql" since we parse GQL schema.

I understand if you don't like macro solution (and can reject pr if you like, seemed like cleanest way I could think of atm)